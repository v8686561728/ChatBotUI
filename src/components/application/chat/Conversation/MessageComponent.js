import React, { useState } from "react";
import PropTypes from "prop-types";
import parse from "html-react-parser";
import { Input } from "antd";
import { CheckOutlined, SendOutlined } from "@ant-design/icons";
import { timeAgo } from "../../../../helpers/date-helper";

const MessageComponent = ({ data, conversation }) => {
  const { handleOptionClick, handleInputSubmit } = data;
  const [readOnly, setReadOnly] = useState(false);
  const [error, setError] = useState("");
  const [value, setValue] = useState("");
  const handleChange = (e) => {
    e.preventDefault();
    setValue(e.target.value);
  };
  const onOptionClick = (conversation, e) => {
    e.preventDefault();
    handleOptionClick(conversation);
  };
  const handleInput = () => {
    conversation.value = value;
    if (validateInput(conversation)) {
      handleInputSubmit(conversation);
      setError("");
      setReadOnly(true);
    }
  };

  // input validation 
  const validateInput = (conversation) => {
    let regexp = "";
    switch (conversation.key) {
      case "email":
        regexp =
          /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;

        if (regexp.test(String(conversation.value).toLowerCase())) {
          return true;
        }
        setError("Enter a valid email");
        return false;
      case "phone":
        regexp = /^\d{10}$/;
        if (regexp.test(conversation.value)) {
          return true;
        }
        setError("Enter a valid phone number");
        return false;
      default:
        return true;
    }
  };

  // based on the component type message component is choosen
  switch (conversation.component) {
    case "button":
      return (
        <div
          className={`button-option ${conversation.id}`}
          name={conversation.id}
          onClick={(e) => onOptionClick(conversation, e)}
        >
          {conversation.text}
        </div>
      );
    case "input":
      return (
        <div className="gx-bubble-block">
          <div className="gx-bubble">
            <div className="gx-input-label">{conversation.text}</div>

            <Input
            data-testid="input"
              onChange={handleChange}
              readOnly={readOnly}
              addonAfter={
                readOnly ? (
                  <CheckOutlined data-testid="save"/>
                ) : (
                  <SendOutlined data-testid="send" onClick={handleInput} />
                )
              }
            ></Input>

            <span className="error" data-testid="error">
              {error}
            </span>
            <div className="gx-time gx-text-muted gx-text-right gx-mt-2">
              {timeAgo(conversation.sentAt)}
            </div>
          </div>
        </div>
      );
    default:
      return (
        <div className="gx-bubble-block">
          <div className="gx-bubble">
            <div className="gx-message">{parse(conversation.text)}</div>
            <div className="gx-time gx-text-muted gx-text-right gx-mt-2">
              {timeAgo(conversation.sentAt)}
            </div>
          </div>
        </div>
      );
  }
};

MessageComponent.propTypes = {
  data: PropTypes.object,
  conversation: PropTypes.object,
};

export default MessageComponent;
