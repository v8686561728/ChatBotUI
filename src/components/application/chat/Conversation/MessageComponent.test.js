import React from "react";
import { render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import MessageComponent from "./MessageComponent";

describe("Message component", () => {
  it("renders correctly", () => {
    const conversation = {
      type: "sent",
      text: "message sent",
      component: "text",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const tree = renderer
      .create(<MessageComponent conversation={conversation} data={data} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("renders Message component without crashing", () => {
    const conversation = {
      type: "sent",
      text: "message sent",
      component: "text",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    render(<MessageComponent conversation={conversation} data={data} />);
  });
  it("Text Message rendered successfully", () => {
    const conversation = {
      type: "sent",
      text: "message sent",
      component: "text",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    const message = component.getByText("message sent", {
      selector: "div",
    });

    expect(message).toBeDefined();
  });
  it("Button Option rendered successfully", () => {
    const conversation = {
      type: "sent",
      text: "Button",
      component: "button",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    const message = component.getByText("Button", {
      selector: "div",
    });

    expect(message).toBeDefined();
  });
  it("Input Option rendered successfully", () => {
    const conversation = {
      type: "sent",
      text: "Email",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    const message = component.getByText("Email");

    expect(message).toBeDefined();
  });
  it("should be able to type in input field", () => {
    const conversation = {
      type: "sent",
      text: "Email",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const { getByTestId } = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    fireEvent.change(getByTestId("input"), {
      target: { value: "test@email.com" },
    });
    expect(getByTestId("input").value).toBe("test@email.com");
  });
  it("Validate the entered email id successful", () => {
    const conversation = {
      type: "sent",
      text: "Email",
      key: "email",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    fireEvent.change(component.getByTestId("input"), {
      target: { value: "test@email.com" },
    });
    fireEvent.click(component.getByTestId("send"));
    expect(component.getByTestId("save"));
  });
  it("Validate the entered email id failure", () => {
    const conversation = {
      type: "sent",
      text: "Email",
      key: "email",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    fireEvent.change(component.getByTestId("input"), {
      target: { value: "test@email" },
    });
    fireEvent.click(component.getByTestId("send"));
    const errorMessage = component.getByText("Enter a valid email", {
      selector: "span",
    });

    expect(errorMessage).toBeDefined();
    expect(component.getByTestId("send"));
  });
  it("Validate the entered phone Number successful", () => {
    const conversation = {
      type: "sent",
      text: "Phone",
      key: "phone",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    fireEvent.change(component.getByTestId("input"), {
      target: { value: "1472583690" },
    });
    fireEvent.click(component.getByTestId("send"));
    expect(component.getByTestId("save"));
  });
  it("Validate the entered phone number failure", () => {
    const conversation = {
      type: "sent",
      text: "Phone",
      key: "phone",
      component: "input",
      sentAt: new Date(),
    };
    const data = {
      handleOptionClick: jest.fn(() => {
        return;
      }),
      handleInputSubmit: jest.fn(() => {
        return;
      }),
    };
    const component = render(
      <MessageComponent conversation={conversation} data={data} />
    );
    fireEvent.change(component.getByTestId("input"), {
      target: { value: "123" },
    });
    fireEvent.click(component.getByTestId("send"));
    const errorMessage = component.getByText("Enter a valid phone number", {
      selector: "span",
    });

    expect(errorMessage).toBeDefined();
    expect(component.getByTestId("send"));
  });
});
