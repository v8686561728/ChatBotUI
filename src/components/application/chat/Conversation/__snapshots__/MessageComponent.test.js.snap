// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message component renders correctly 1`] = `
<div
  className="gx-bubble-block"
>
  <div
    className="gx-bubble"
  >
    <div
      className="gx-message"
    >
      message sent
    </div>
    <div
      className="gx-time gx-text-muted gx-text-right gx-mt-2"
    >
      now
    </div>
  </div>
</div>
`;
